## Import Libraries

Importing python libraries to be used in this data analysis:

* **sqlite3** for interacting with local relational database.
* **numpy** and **pandas** for data ingestion and manipulation.
* **matplotlib** for data visualization.
* **sklearn** for Machine Learning.
* **customplot** which contains custom functions written for this notebook.

In [19]:
import sqlite3
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
from sklearn.preprocessing import scale
# from customplot import *

<h1 style="font-size:2em;color:#2467C0">Data Acquisition</h1>

Command below read the dataset
Make sure the run the import cell above(shift + enter) before running ingest code below.

**df** is a variable pointing to a pandas data frame.

Remember to have the database and the jupyter notebook in the same folder

In [20]:
# Create your connection
con = sqlite3.connect('database.sqlite')
df = pd.read_sql_query("SELECT * FROM Player_Attributes", con)

<h1 style="font-size:2em;color:#2467C0">Exploring Data</h1

Data exploration by generating simple statistics of the data

Checking data columns using a pandas attribute called "columns"

In [21]:
df.columns

Index(['id', 'player_fifa_api_id', 'player_api_id', 'date', 'overall_rating',
       'potential', 'preferred_foot', 'attacking_work_rate',
       'defensive_work_rate', 'crossing', 'finishing', 'heading_accuracy',
       'short_passing', 'volleys', 'dribbling', 'curve', 'free_kick_accuracy',
       'long_passing', 'ball_control', 'acceleration', 'sprint_speed',
       'agility', 'reactions', 'balance', 'shot_power', 'jumping', 'stamina',
       'strength', 'long_shots', 'aggression', 'interceptions', 'positioning',
       'vision', 'penalties', 'marking', 'standing_tackle', 'sliding_tackle',
       'gk_diving', 'gk_handling', 'gk_kicking', 'gk_positioning',
       'gk_reflexes'],
      dtype='object')

Displaying simple statistics of the dataset. Run each cell to make sure you see the outputs.

In [22]:
df.describe().transpose()

count           mean            std     min        25%  \
id                  183978.0   91989.500000   53110.018250     1.0   45995.25   
player_fifa_api_id  183978.0  165671.524291   53851.094769     2.0  155798.00   
player_api_id       183978.0  135900.617324  136927.840510  2625.0   34763.00   
overall_rating      183142.0      68.600015       7.041139    33.0      64.00   
potential           183142.0      73.460353       6.592271    39.0      69.00   
crossing            183142.0      55.086883      17.242135     1.0      45.00   
finishing           183142.0      49.921078      19.038705     1.0      34.00   
heading_accuracy    183142.0      57.266023      16.488905     1.0      49.00   
short_passing       183142.0      62.429672      14.194068     3.0      57.00   
volleys             181265.0      49.468436      18.256618     1.0      35.00   
dribbling           183142.0      59.175154      17.744688     1.0      52.00   
curve               181265.0      52.965675      18.255788     2.0      41.00   
free_kick_accuracy  183142.0      49.380950      17.831746     1.0      36.00   
long_passing        183142.0      57.069880      14.394464     3.0      49.00   
ball_control        183142.0      63.388879      15.196671     5.0      58.00   
acceleration        183142.0      67.659357      12.983326    10.0      61.00   
sprint_speed        183142.0      68.051244      12.569721    12.0      62.00   
agility             181265.0      65.970910      12.954585    11.0      58.00   
reactions           183142.0      66.103706       9.155408    17.0      61.00   
balance             181265.0      65.189496      13.063188    12.0      58.00   
shot_power          183142.0      61.808427      16.135143     2.0      54.00   
jumping             181265.0      66.969045      11.006734    14.0      60.00   
stamina             183142.0      67.038544      13.165262    10.0      61.00   
strength            183142.0      67.424529      12.072280    10.0      60.00   
long_shots          183142.0      53.339431      18.367025     1.0      41.00   
aggression          183142.0      60.948046      16.089521     6.0      51.00   
interceptions       183142.0      52.009271      19.450133     1.0      34.00   
positioning         183142.0      55.786504      18.448292     2.0      45.00   
vision              181265.0      57.873550      15.144086     1.0      49.00   
penalties           183142.0      55.003986      15.546519     2.0      45.00   
marking             183142.0      46.772242      21.227667     1.0      25.00   
standing_tackle     183142.0      50.351257      21.483706     1.0      29.00   
sliding_tackle      181265.0      48.001462      21.598778     2.0      25.00   
gk_diving           183142.0      14.704393      16.865467     1.0       7.00   
gk_handling         183142.0      16.063612      15.867382     1.0       8.00   
gk_kicking          183142.0      20.998362      21.452980     1.0       8.00   
gk_positioning      183142.0      16.132154      16.099175     1.0       8.00   
gk_reflexes         183142.0      16.441439      17.198155     1.0       8.00   

                         50%        75%       max  
id                   91989.5  137983.75  183978.0  
player_fifa_api_id  183488.0  199848.00  234141.0  
player_api_id        77741.0  191080.00  750584.0  
overall_rating          69.0      73.00      94.0  
potential               74.0      78.00      97.0  
crossing                59.0      68.00      95.0  
finishing               53.0      65.00      97.0  
heading_accuracy        60.0      68.00      98.0  
short_passing           65.0      72.00      97.0  
volleys                 52.0      64.00      93.0  
dribbling               64.0      72.00      97.0  
curve                   56.0      67.00      94.0  
free_kick_accuracy      50.0      63.00      97.0  
long_passing            59.0      67.00      97.0  
ball_control            67.0      73.00      97.0  
acceleration            69.0   

<h1 style="font-size:2em;color:#2467C0">Data Cleaning: Handling Missing Data</h1>

Clean data by converting or getting rid of null or missing values. The next code cell shows if any of the 183978 rows have null value in one of the 42 columns

In [23]:
# Is any row NULL?
df.isnull().any().any(), df.shape

(True, (183978, 42))

Next code cell tries to find out how many data points in each column are null.

In [24]:
df.isnull().sum(axis=0)

id                        0
player_fifa_api_id        0
player_api_id             0
date                      0
overall_rating          836
potential               836
preferred_foot          836
attacking_work_rate    3230
defensive_work_rate     836
crossing                836
finishing               836
heading_accuracy        836
short_passing           836
volleys                2713
dribbling               836
curve                  2713
free_kick_accuracy      836
long_passing            836
ball_control            836
acceleration            836
sprint_speed            836
agility                2713
reactions               836
balance                2713
shot_power              836
jumping                2713
stamina                 836
strength                836
long_shots              836
aggression              836
interceptions           836
positioning             836
vision                 2713
penalties               836
marking                 836
standing_tackle     

# Fixing Null Values By Deleting Them

The next two lines drops the null values by going through each row.

In [25]:
# Take initial number of rows
rows = df.shape[0]

# Drop the NULL values
df = df.dropna()

In [26]:
# Check if all NULLL values are gone?
print(rows)
df.isnull().any().any(), df.shape

183978


(False, (180354, 42))

In [27]:
# How many rows with NULL values?
rows - df.shape[0]

3624

The rows can be randomly shuffled so as not to have the same 

In [28]:
#Shuffle the rows of df so we get a distributed sample when we display top few rows

df = df.reindex(np.random.permutation(df.index))

<h1 style="font-size:2em;color:#2467C0">Predicting: 'Overall_Rating' of a player</h1>

The data is now clean giving room to better understanding and diving further into the dataset.

To take a look at the first few rows of the dataset; use the head function for data frames for this task.

This gives every column in every row.

In [29]:
df.head(5)

id  player_fifa_api_id  player_api_id                 date  \
181956  181957              185181         192064  2010-02-22 00:00:00   
98558    98559              201332         188503  2015-04-24 00:00:00   
165055  165056              187043          70078  2013-08-23 00:00:00   
100583  100584              205976         358127  2014-07-18 00:00:00   
76051    76052              186130         163200  2016-04-14 00:00:00   

        overall_rating  potential preferred_foot attacking_work_rate  \
181956            61.0       74.0          right              medium   
98558             57.0       62.0          right              medium   
165055            73.0       80.0           left              medium   
100583            75.0       80.0          right                high   
76051             72.0       73.0          right              medium   

       defensive_work_rate  crossing     ...       vision  penalties  marking  \
181956              medium      21.0     ...         54.0       51.0     21.0   
98558               medium      52.0     ...         61.0       57.0     50.0   
165055              medium      58.0     ...         73.0       42.0     58.0   
100583                high      60.0     ...         65.0       64.0     25.0   
76051               medium      54.0     ...         49.0       52.0     73.0   

        standing_tackle  sliding_tackle  gk_diving  gk_handling  gk_kicking  \
181956             21.0            11.0       60.0         61.0        60.0   
98558              54.0            47.0        8.0          6.0         7.0   
165055             54.0            42.0        6.0         15.0         6.0   
100583             23.0            25.0        7.0         13.0         8.0   
76051              76.0            76.0       11.0         12.0        12.0   

        gk_positioning  gk_reflexes  
181956            57.0         65.0  
98558              7.0          9.0  
165055             8.0         13.0  
100583             5.0          8.0  
76051             14.0         13.0  

[5 rows x 42 columns]

In the case you are interested in plotting some columns, use pandas column selection option as follows. Ignore the first column in the output of the one line code below. It is the unique identifier that acts as an index for the data.

In [30]:
df[:10][['penalties', 'overall_rating']]

penalties  overall_rating
181956       51.0            61.0
98558        57.0            57.0
165055       42.0            73.0
100583       64.0            75.0
76051        52.0            72.0
147054       11.0            69.0
101969       53.0            54.0
84062        49.0            75.0
78441        62.0            65.0
79695        42.0            56.0

### Feature Correlation Analysis

Check if "penalties" is correlated to "overall_rating". To do this, use a similar selection operation, but for all rows and within the correlation function.

# Pearson's Correlation Coefficient


In [31]:
# Are these correlated ?
df['overall_rating'].corr(df['penalties'])

0.3927151079111884

Pearson's Correlation Coefficient for the two columns is 0.39

Pearson goes from -1 to +1. A value of 0 shows no correlation. A value greater than 0 shows a positive correlation; as the value of one variable increases, so does the other variable. A value less than 0 indicates a negative correlation; as the value of one variable increases, the value of the other variable decreases.

A value of 0.39 shows some correlation but it could be stronger.

Now that the attributes are slightly correlated, it gives hope that a meaningful predictor can be built using this 'weakly' correlated features.

Now, create a list of features that will iterate the same operation on.

## Create a list of potential Features that you want to measure correlation with

In [32]:
potentialFeatures = ['acceleration', 'curve', 'free_kick_accuracy', 'ball_control', 'shot_power', 'stamina']

The for loop below prints out the correlation coeeficient of "overall_rating" of a player with each feature added to the list as potential

In [33]:
for f in potentialFeatures:
    related = df['overall_rating'].corr(df[f])
    print("%s: %f" %(f, related))

acceleration: 0.243998
curve: 0.357566
free_kick_accuracy: 0.349800
ball_control: 0.443991
shot_power: 0.428053
stamina: 0.325606


**"ball_control"(0.44)** and **"shot_power"(0.43)** are features that have the highest correlation coefficient with "overall_rating"